src/hooks/proxy.js

```diff
@@ -36,12 +66,21 @@ function validateProxyProtocol(protocol) {
     }
 }
 
+exports.agentCache = new Map();
```

Comment:
Looks like you never clear this map except in tests, do sounds like a memory leak to me. Maybe it could be a WeakMap instead so we dont have to care about clearing it?

Reply:
https://github.com/apify/got-scraping/blob/a179224748e3626b194a745d5d5a4e4e5617e561/test/proxy.test.js#L26-L31

Reply:
Yes, that is in tests, how do you clear it in real world usage? Manually?

Reply:
> Maybe it could be a WeakMap instead so we dont have to care about clearing it?

It needs be cleared only when the conditions change. E.g. the expected proxy server is not `h2` anymore but `http/1.1`. Also `WeakMap` accepts objects, while we need strings.

Reply:
> Yes, that is in tests, how do you clear it in real world usage? Manually?

Unless the proxy url changes continuously, there's nothing wrong. In this case we can use https://github.com/sindresorhus/quick-lru ?

Reply:
We could use `WeakRef` but that would require us to bump the required Node.js version to 14.

Reply:
The required Node version is 15, no?

```
"engines": {
        "node": ">=15.10.0"
    },
```

Reply:
Ah great, I thought it was `10` as the previous code implied this :P

After all `WeakRef` may safe as it may gc in undesirable moments. E.g. when no request is being sent for a while. Then `agent` has no references (active sockets), so the object would get removed.

Are you ok with `quick-lru`?

Reply:
> Ah great, I thought it was 10 as the previous code implied this :P

Yeah, we probably forgot some comments in the codebase when we were releasing 2.0.

Regarding the `quick-lru`, I'm ok with it.

Reply:
Added `quick-lru`. PTAL :)